pysat/utils/io.py

```diff
@@ -0,0 +1,1029 @@
+                                  'Format': inst._get_var_type_code(scoltype),
+                                  'Var_Type': 'data'}
+                    inst.meta[svar] = smeta_dict
+            else:
+                meta_dict['Format'] = inst._get_var_type_code(coltype)
+
+            # Update the meta data
+            inst.meta[var] = meta_dict
+        else:
+            pysat.logger.info(''.join(('Unable to find MetaData for ', var)))
+
+    return
+
+
+def load_netcdf4(fnames, strict_meta=False, file_format=None,
+                 epoch_name='Epoch', pandas_format=True,
+                 labels={'units': ('units', str), 'name': ('long_name', str),
+                         'notes': ('notes', str), 'desc': ('desc', str),
+                         'plot': ('plot_label', str), 'axis': ('axis', str),
+                         'scale': ('scale', str),
+                         'min_val': ('value_min', np.float64),
+                         'max_val': ('value_max', np.float64),
+                         'fill_val': ('fill', np.float64)}):
+    """Load netCDF-3/4 file produced by pysat.
+
+    Parameters
+    ----------
+    fnames : str, array_like, or NoneType
+        Filename(s) to load, will fail if None (default=None)
+    strict_meta : bool
+        Flag that checks if metadata across fnames is the same if True
+        (default=False)
+    file_format : str or NoneType
+        file_format keyword passed to netCDF4 routine.  Expects one of
+        'NETCDF3_CLASSIC', 'NETCDF3_64BIT', 'NETCDF4_CLASSIC', or 'NETCDF4'.
+        If None, defaults to 'NETCDF4'. (default=None)
+    epoch_name : str
+        Data key for time variable (default='Epoch')
+    pandas_format : bool
+        Flag specifying if data is stored in a pandas DataFrame (True) or
+        xarray Dataset (False). (default=False)
+    labels : dict
+        Dict where keys are the label attribute names and the values are tuples
+        that have the label values and value types in that order.
+        (default={'units': ('units', str), 'name': ('long_name', str),
+        'notes': ('notes', str), 'desc': ('desc', str),
+        'min_val': ('value_min', np.float64),
+        'max_val': ('value_max', np.float64), 'fill_val': ('fill', np.float64)})
+
+    Returns
+    -------
+    out : pandas.DataFrame or xarray.Dataset
+        Class holding file data
+    meta : pysat.Meta
+        Class holding file meta data
+
+    Raises
+    ------
+    KeyError
+        If epoch/time dimension could not be identified.
+
+    """
+    # Process the input values
+    fnames = pysat.utils.listify(fnames)
+
+    if file_format is None:
+        file_format = 'NETCDF4'
+    else:
+        file_format = file_format.upper()
+
+    # Initialize local variables
+    saved_meta = None
+    running_idx = 0
+    running_store = []
+    two_d_keys = []
+    two_d_dims = []
+    meta = pysat.Meta(labels=labels)
+
+    # Load data by type
+    if pandas_format:
```

Comment:
To reduce complexity, we could add two hidden functions: `_load_netcdf_pandas` and `_load_netcdf_xarray` and then just choose which to run here.  The nesting in the pandas section is already getting fairly deep, and this would buy back some space as well as help when debugging for specific files.
```suggestion
    if pandas_format:
        out, meta = _load_netcdf_pandas(fname, format, etc)
    else:
        out, meta = _load_netcdf_xarray(fname, format, etc)
```

Reply:
Lines below would need to be moved into the new functions

Reply:
I am ok doing this.

Reply:
Done!